### PR TITLE
SONARPY-1711: Rule S5659: do not raise under certain use of get_unverified_header()

### PR DESCRIPTION
**Pull request by joke1196**

### CODE REVIEW

File: python-checks/src/main/java/org/sonar/python/checks/JwtVerificationCheck.java [hunk 1]
```diff
@@ -175,4 +186,83 @@ private static boolean isCallToVerifyJwt(Tree t) {
...
+  private static Stream<StringLiteral> argumentsAccessedByGet(Symbol symbol, CallExpression call) {
+    var usages = getHeaderDictUsages(symbol, call);
+    var parentOfUsages = usages.map(Usage::tree).map(Tree::parent);
+    var callExpressionsFromUsages = getCallExprWhereDictIsAccessedWithGet(parentOfUsages);
+    var callExpressionFromGetUnverifiedHeaders = getCallExprWhereDictIsAccessedWithGet(Stream.of(call.parent()));
+    var arguments = Stream.concat(callExpressionsFromUsages, callExpressionFromGetUnverifiedHeaders).map(CallExpression::arguments).flatMap(Collection::stream);
+    return getStringLiteralArguments(arguments);
+  }
+
+  private static Stream<Usage> getHeaderDictUsages(Symbol symbol, CallExpression call) {
```
Comment (Seppli11) [inline python-checks/src/main/java/org/sonar/python/checks/JwtVerificationCheck.java:222]: I think a name, which indicates that only usages after the definition are returned, might be better. E.g. `getForwardUsages`

File: python-checks/src/main/java/org/sonar/python/checks/JwtVerificationCheck.java [hunk 2]
```diff
@@ -175,4 +186,83 @@ private static boolean isCallToVerifyJwt(Tree t) {
...
+    return symbol.usages().stream()
+      .filter(usage -> usage.tree().firstToken().line() > call.callee().firstToken().line());
+  }
+
+  private static Stream<CallExpression> getCallExprWhereDictIsAccessedWithGet(Stream<Tree> parentQualifiedExpr) {
+    return parentQualifiedExpr
+      .filter(parent -> parent.is(Tree.Kind.QUALIFIED_EXPR))
+      .map(TreeUtils.toInstanceOfMapper(QualifiedExpression.class))
```
Comment (Seppli11) [inline python-checks/src/main/java/org/sonar/python/checks/JwtVerificationCheck.java:230]: Can be replaced with the following:
```suggestion
      .flatMap(TreeUtils.toStreamInstanceOfMapper(QualifiedExpression.class))
```

File: python-checks/src/main/java/org/sonar/python/checks/JwtVerificationCheck.java [hunk 3]
```diff
@@ -175,4 +186,83 @@ private static boolean isCallToVerifyJwt(Tree t) {
...
+      .filter(Objects::nonNull)
+      .filter(expr -> "get".equals(expr.name().name()))
+      .filter(expr -> expr.parent().is(Kind.CALL_EXPR))
+      .map(QualifiedExpression::parent)
+      .map(TreeUtils.toInstanceOfMapper(CallExpression.class));
```
Comment (Seppli11) [inline python-checks/src/main/java/org/sonar/python/checks/JwtVerificationCheck.java:235]: Could be written as
```suggestion
      .map(QualifiedExpression::parent)
      .flatMap(TreeUtils.toStreamInstanceOfMapper(CallExpression.class));
```

File: python-checks/src/main/java/org/sonar/python/checks/JwtVerificationCheck.java [hunk 4]
```diff
@@ -175,4 +186,83 @@ private static boolean isCallToVerifyJwt(Tree t) {
...
+  }
+
+  private static Stream<StringLiteral> getStringLiteralArguments(Stream<Argument> arguments) {
+    return arguments.filter(arg -> arg.is(Tree.Kind.REGULAR_ARGUMENT))
+      .map(TreeUtils.toInstanceOfMapper(RegularArgument.class))
```
Comment (Seppli11) [inline python-checks/src/main/java/org/sonar/python/checks/JwtVerificationCheck.java:240]:
```suggestion
    return arguments
      .flatMap(TreeUtils.toStreamInstanceOfMapper(RegularArgument.class))
```

File: python-checks/src/main/java/org/sonar/python/checks/JwtVerificationCheck.java [hunk 5]
```diff
@@ -175,4 +186,83 @@ private static boolean isCallToVerifyJwt(Tree t) {
...
+      .map(RegularArgument::expression)
+      .map(TreeUtils.toInstanceOfMapper(StringLiteral.class))
+      .filter(Objects::nonNull);
```
Comment (Seppli11) [inline python-checks/src/main/java/org/sonar/python/checks/JwtVerificationCheck.java:243]:
```suggestion
      .flatMap(TreeUtils.toStreamInstanceOfMapper(StringLiteral.class));
```

**Review by Seppli11 — CHANGES_REQUESTED** [review]
Is it on purpose that something like `foo(jwt.get_unverified_header(token).get("x5u"))` or `return jwt.get_unverified_header(token).get("x5u")` is not compliant? 
If yes, I think, a test case should be added to document this.  

Apart from this I've made a comment when `TreeUtils.toStreamInstanceOfMapper` can be used instead of first filtering and then mapping.

Otherwise, it looks good

EDIT: fixed the second example

**Comment by joke1196** [comment]
So the first is not specified in the ticket but I would think it should be compliant, I will fix this. The second is documented in the ticket and in test cases as the token passed to the function does not affect how we would raise issues.

**Review by Seppli11 — APPROVED** [review]
lgtm!

**Pull request merged**